### ISSUE TITLE
Local variables being overwritten when realizing lazy seqs

### ISSUE THREAD

**Issue by n-hansen** [babashka/babashka#1006]
**version**
0.6.0

**platform**
Mac OS X

**problem**

If a lazy seq is constructed via `for` within a lambda that gets called repeatedly before the seq is fully realized, any captured variables get rebound for _all_ produced seqs.

**repro**
```clj
#!/usr/bin/env bb
;; example.bb
(let [f (fn [x] (for [_ (range 2)
                      _ (range 1)]
                  x))
      s1 (f :a)
      s2 (f :b)]
  (first s2) ; realize the head of s2
  (println (vec s1)))
```

```sh
$ ./example.bb
[:a :b]
```

**expected behavior**

The function's parameters should remain distinct for each function call. If the example is executed in a clojure or clojurescript repl, the output is `[:a :a]` as expected.

**Comment by borkdude** [comment]
I agree, this is a bug. Will look into it soon, thanks for reporting.

**Comment by borkdude** [comment]
This is the full expansion of a repro. I hope to get a much smaller repro though, as I don't see yet what's going on.

```
(defn foo [x]
  (clojure.core/let
      [iter__2205__auto__
       (clojure.core/fn
         iter__9965
         [s__9966]
         (clojure.core/lazy-seq
          (loop
              [s__9966 s__9966]
            (clojure.core/when-first
                [_ s__9966]
              (clojure.core/let
                  [iterys__2201__auto__
                   (clojure.core/fn
                     iter__9967
                     [s__9968]
                     (clojure.core/lazy-seq
                      (loop
                          [s__9968 s__9968]
                        (clojure.core/let
                            [s__9968 (clojure.core/seq s__9968)]
                          (clojure.core/when-let
                              [s__9968 (clojure.core/seq s__9968)]
                            (if
                                (clojure.core/chunked-seq? s__9968)
                              (clojure.core/let
                                  [c__2203__auto__
                                   (clojure.core/chunk-first s__9968)
                                   size__2204__auto__
                                   (clojure.core/int (clojure.core/count c__2203__auto__))
                                   b__9970
                                   (clojure.core/chunk-buffer size__2204__auto__)]
                                (if
                                    (loop
                                        [i__9969 (clojure.core/int 0)]
                                      (if
                                          (clojure.core/< i__9969 size__2204__auto__)
                                        (clojure.core/let
                                            [_ (clojure.core/nth c__2203__auto__ i__9969)]
                                          (do
                                            (clojure.core/chunk-append b__9970 x)
                                            (recur (clojure.core/unchecked-inc i__9969))))
                                        true))
                                  (clojure.core/chunk-cons
                                   (clojure.core/chunk b__9970)
                                   (iter__9967 (clojure.core/chunk-rest s__9968)))
                                  (clojure.core/chunk-cons
                                   (clojure.core/chunk b__9970)
                                   nil)))
                              (clojure.core/let
                                  [_ (clojure.core/first s__9968)]
                                (clojure.core/cons
                                 x
                                 (iter__9967 (clojure.core/rest s__9968))))))))))
                   fs__2202__auto__
                   (clojure.core/seq (iterys__2201__auto__ (range 2)))]
                (if
                    fs__2202__auto__
                  (clojure.core/concat
                   fs__2202__auto__
                   (iter__9965 (clojure.core/rest s__9966)))
                  (recur (clojure.core/rest s__9966))))))))]
    (iter__2205__auto__ (range 2))))

(def xs (foo :x))
(def ys (foo :y))
xs
```

**Comment by borkdude** [comment]
This used to work with bb 0.4.0 but I can reproduce it with 0.5.0 so somewhere in between there must have happened a change which broke this.

**Comment by borkdude** [comment]
I found the SCI commit which introduced this regression:

https://github.com/borkdude/sci/commit/9f16477b8678c98878dde192b96aae916085b7ab

**Comment by borkdude** [comment]
I have found the issue and working on a solution.

**Comment by borkdude** [comment]
Here is the shortest repro I can come up with:

```
(defn foof [x]
  (let [f (fn f
            ([] (f nil))
            ([_] x))]
    f))

(def f1 (foof :a))
(def f2 (foof :b))

(f1)
``` 

The issue has to do with self-reference of a function.

**Comment by borkdude** [comment]
@n-hansen The issue should be fixed in this version:

https://22356-201467090-gh.circle-artifacts.com/0/release/babashka-0.6.1-SNAPSHOT-macos-amd64.tar.gz

I would appreciate it if you could verify this.

**Comment by n-hansen** [comment]
I'm unable to reproduce the issue with the 0.6.1 snapshot. Thanks for the quick turnaround!

**Comment by borkdude** [comment]
@n-hansen Released in 0.6.1.
